src/rust/engine/src/rule_graph.rs

```diff
@@ -0,0 +1,583 @@
+      value: None,
+      rule: None,
+      selector: Some(selector),
+      reason: None,
+    }
+  }
+
+  fn new_inner(subject_type: TypeId, rule: &Task) -> Entry {
+    Entry {
+      entry_type: EntryType::InnerEntry,
+      subject_type: Some(subject_type),
+      value: None,
+      rule: Some(rule.clone()), // TODO decide on clone vs lifetimes
+      selector: None,
+      reason: None,
+    }
+  }
+
+  fn new_subject_is_product(subject_type: TypeId) -> Entry {
+    Entry {
+      entry_type: EntryType::SubjectIsProduct,
+      subject_type: Some(subject_type),
+      value: None,
+      rule: None,
+      selector: None,
+      reason: None,
+    }
+  }
+
+  fn new_unreachable(rule: &Task) -> Entry {
+    Entry {
+      entry_type: EntryType::Unreachable,
+      subject_type: None,
+      value: None,
+      rule: Some(rule.clone()), // TODO decide on clone vs lifetimes
+      selector: None,
+      reason: None,
+    }
+  }
+  fn new_literal(value: Key, product: TypeConstraint) -> Entry {
+    Entry {
+      entry_type: EntryType::Literal,
+      subject_type: None,
+      value: Some(value),
+      rule: None,
+      selector: Some(product),
+      reason: None,
+    }
+  }
+
+  fn subject_type(&self) -> TypeId {
+    if let Some(subject_type) = self.subject_type {
+      subject_type
+    } else {
+      panic!("no subject type")
+    }
+  }
+
+  fn rule(&self) -> &Task {
+    if let Some(ref rule) = self.rule {
+      rule
+    } else {
+      panic!("no rule")
+    }
+  }
+  fn input_selectors(&self) -> Vec<Selector> {
+    match self.entry_type {
+      EntryType::InnerEntry =>
+        match self.rule {
+          Some(ref task) => {
+            task.clause.clone()
+          },
+          None => vec![] // Could also panic
+        },
+      EntryType::Root => match self.selector {
+        Some(s) => vec![Selector::Select(Select { product: s, variant_key: None })],
+        None => vec![]
+      },
+      _ => panic!("this entry doesn't have selectors")
+    }
+  }
+
+  fn can_have_dependencies(&self) -> bool {
+    match self.entry_type {
+      EntryType::SubjectIsProduct => false,
+      EntryType::Literal => false,
+      EntryType::InnerEntry => true,
+      EntryType::Root => true,
+      EntryType::Unreachable => false,
+    }
+  }
+  fn can_be_dependency(&self) -> bool {
+    match self.entry_type {
+      EntryType::SubjectIsProduct => true,
+      EntryType::Literal => true,
+      EntryType::InnerEntry => true,
+      EntryType::Root => false,
+      EntryType::Unreachable => false,
+    }
+  }
+}
+
+type Entries = Vec<Entry>;
+type RootRuleDependencyEdges = HashMap<Entry, RuleEdges>; // Root -> InternalEntry
+type Rule = Task;
+type RuleDependencyEdges = HashMap<Entry, RuleEdges>; // InternalEntry -> InternalEntry
+type RuleDiagnostics = Vec<Diagnostic>;
+type UnfulfillableRuleMap = HashMap<Entry, RuleDiagnostics>;
+
+#[derive(Debug)]
+pub struct RootSubjectTypes<'a> {
+  pub subject_types: &'a Vec<TypeId>
```

Comment:
`TypeId`s are tiny, so should almost certainly just clone rather than tracking lifetimes.